src/main/java/de/tum/in/www1/artemis/web/rest/BuildPlanResource.java

```diff
@@ -51,4 +65,50 @@ public ResponseEntity<String> getBuildPlan(@PathVariable Long exerciseId, @Reque
 
         return ResponseEntity.ok().body(buildPlan.getBuildPlan());
     }
+
+    /**
+     * Returns the build plan for a given programming exercise.
+     *
+     * @param exerciseId the exercise for which the build plan should be retrieved
+     * @return the build plan stored in the database
+     */
+    @GetMapping(BUILD_PLAN + "/for-editor")
+    @PreAuthorize("permitAll()")
+    public ResponseEntity<BuildPlan> getBuildPlanForEditor(@PathVariable Long exerciseId) {
+        log.debug("REST request to get build plan for programming exercise with id {}", exerciseId);
+
+        final BuildPlan buildPlan = buildPlanRepository.findByProgrammingExercises_IdWithProgrammingExercisesElseThrow(exerciseId);
+        // orElseThrow is safe here since the query above ensures that we find a build plan that is attached to that exercise
+        final ProgrammingExercise programmingExercise = buildPlan.getProgrammingExerciseById(exerciseId)
+                .orElseThrow(() -> new EntityNotFoundException("Could not find connected exercise for build plan."));
+
+        authorizationCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.EDITOR, programmingExercise, null);
+
+        return ResponseEntity.ok().body(buildPlan);
+    }
+
+    /**
+     * Updates the build plan for the given exercise.
+     * <p>
+     * Triggers a template and solution build to give feedback if the new build plan works as expected.
+     *
+     * @param exerciseId The exercise for which the build plan should be updated.
+     * @param buildPlan  The new build plan for the exercise.
+     * @return The updated build plan.
+     */
+    @PutMapping("/programming-exercises/{exerciseId}/build-plan")
+    @PreAuthorize("hasRole('EDITOR')")
+    public ResponseEntity<BuildPlan> setBuildPlan(@PathVariable Long exerciseId, @RequestBody BuildPlan buildPlan) {
+        log.debug("REST request to set build plan for programming exercise with id {}", exerciseId);
+
+        final ProgrammingExercise programmingExercise = programmingExerciseRepository.findByIdElseThrow(exerciseId);
```

Comment:
Please also check that buildPlan.exercise.id is equal to the exerciseId.

Reply:
The build plan object from the request body here actually only contains the String, the set of programming exercises is not considered. While the programming exercise would be accessible in the front-end, it is omitted to keep the data transfer lightweight.